llvm/lib/Target/EVM/EVMISelLowering.cpp

```diff
@@ -758,3 +761,135 @@ MachineBasicBlock *EVMTargetLowering::emitSelect(MachineInstr &MI,
   MI.eraseFromParent(); // The pseudo instruction is gone now.
   return BB;
 }
+
+// Fold
+//   (select C, (op Y, X), Y)  ->  (op Y, (select C, X, NEUTRAL)),
+//   when 0 is the neutral element for op.
+//
+// That lets a later combine turn  (select C, X, 0)  into  (C * X).
+//
+// Examples:
+//   (select C, (sub Y, X), Y)  ->  (sub Y, (select C, X, 0))
+//   (select C, (or  Y, X), Y)  ->  (or  Y, (select C, X, 0))
+//   (select C, (or  X, Y), Y)  ->  (or  Y, (select C, X, 0))  ; commutativity
+//   handled.
+static SDValue tryFoldSelectIntoOp(SDNode *N, SelectionDAG &DAG, SDValue TrueV,
+                                   SDValue FalseV, bool Swapped) {
+  bool Commutative = true;
+  unsigned Opc = TrueV.getOpcode();
+  switch (Opc) {
+  default:
+    return SDValue();
+  case ISD::SHL:
+  case ISD::SRA:
+  case ISD::SRL:
+  case ISD::SUB:
+    Commutative = false;
+    break;
+  case ISD::ADD:
+  case ISD::OR:
+  case ISD::XOR:
+    break;
+  }
+
+  if (!TrueV.hasOneUse() || isa<ConstantSDNode>(FalseV))
+    return SDValue();
+
+  unsigned OpToFold = 0;
+  if (FalseV == TrueV.getOperand(0))
+    ;
+  else if (Commutative && FalseV == TrueV.getOperand(1))
+    OpToFold = 1;
+  else
+    return SDValue();
+
+  EVT VT = N->getValueType(0);
+  SDLoc DL(N);
+  SDValue OtherOp = TrueV.getOperand(1 - OpToFold);
+  EVT OtherOpVT = OtherOp.getValueType();
+  SDValue IdentityOperand =
+      DAG.getNeutralElement(Opc, DL, OtherOpVT, N->getFlags());
+  if (!Commutative)
+    IdentityOperand = DAG.getConstant(0, DL, OtherOpVT);
+  assert(IdentityOperand && "No identity operand!");
```

Comment:
```suggestion
  SDValue IdentityOperand = DAG.getConstant(0, DL, OtherOpVT);
```
No need to call `DAG.getNeutralElement` since 0 is neutral element for all above operations.